src/Facilitator/Facilitator.js

```diff
@@ -662,6 +662,136 @@ class Facilitator {
       });
   }
 
+  /**
+   * Perform redeem.
+   *
+   * @param {string} amount Redeem amount
+   * @param {string} beneficiary Beneficiary address.
+   * @param {string} gasPrice Gas price.
+   * @param {string} gasLimit Gas limit.
+   * @param {string} nonce Redeemer's nonce.
+   * @param {string} hashLock Hash lock;
+   * @param {Object} txOptions Transaction options.
+   * @return {Promise} Promise object
+   */
+  async redeem(amount, beneficiary, gasPrice, gasLimit, hashLock, txOptions) {
+    logger.info('Redeem');
+    logger.info('-----------------------');
+    if (new BN(amount).eqn(0)) {
+      const err = new TypeError('Redeem amount must not be zero.');
+      return Promise.reject(err);
+    }
+
+    if (!Web3.utils.isAddress(beneficiary)) {
+      const err = new TypeError('Invalid beneficiary address.');
+      return Promise.reject(err);
+    }
+
+    if (typeof gasPrice !== 'string') {
+      const err = new TypeError('Invalid gas price.');
+      return Promise.reject(err);
+    }
+
+    if (typeof gasLimit !== 'string') {
+      const err = new TypeError('Invalid gas limit.');
+      return Promise.reject(err);
+    }
+
+    if (typeof hashLock !== 'string') {
+      const err = new TypeError('Invalid hash lock.');
+      return Promise.reject(err);
+    }
+
+    if (!txOptions) {
+      const err = new TypeError('Invalid transaction options.');
+      return Promise.reject(err);
+    }
+
+    if (!Web3.utils.isAddress(txOptions.from)) {
+      const err = new TypeError('Invalid redeemer address.');
+      return Promise.reject(err);
+    }
+
+    const redeemer = txOptions.from;
+
+    logger.info('* Getting bounty amount *');
+    const bounty = await this.coGateway.getBounty().catch((exception) => {
+      logger.info('  - Exception while getting bounty amount');
+      return Promise.reject(exception);
+    });
+
+    if (!new BN(txOptions.value).eq(new BN(bounty))) {
+      logger.info(
+        '  - Value in transaction option is not equal to the bounty amount',
+      );
+      return Promise.reject(false);
+    }
+
+    logger.info(
+      '* Checking if redeemer has approved CoGateway for token transfer *',
+    );
+
+    const isRedeemAmountApproved = await this.coGateway
+      .isRedeemAmountApproved(redeemer, amount)
+      .catch((exception) => {
+        logger.info('  - Exception while checking redeem amount approval');
+        return Promise.reject(exception);
+      });
+
+    logger.info(`  - Approval status is ${isRedeemAmountApproved}`);
+
+    if (!isRedeemAmountApproved) {
+      /*
+       * If a redeemer is initiating the redeem process through this,
+       * then redeemer is the facilitator and its correct to assume that approval
+       * can be done.
+       */
+      logger.info(
+        '  - As Redeemer is facilitator, approving CoGateway for token transfer',
```

Comment:
Is this an assumption that is always true? What if the staker forgot to approve?

Reply:
As per the contract, stake and redeem methods can be called only by staker and redeemer.

Reply:
But how do you know it is the facilitator if it is not approved? Can it not be non-approved for other reasons? 🤔

Reply:
The facilitator is doing stake/redeem, and signing the transaction. So the intention is to move the tokens.

Reply:
I understand the "happy" path. I don't understand how you rule out that something went wrong.

Reply:
Maybe I am missing. Can you give an example?

Reply:
Hm, let me think 🤔 

Let's say Alice wants to redeem 100 USC tokens from the Unsplash chain. KIT is informed by the system. For some reason, there is no approval of Alice's USC to the gateway. Or maybe it is too low.
Now KIT (or another worker) pick up her redeem and execute it.
▶️The redemption amount is now taken from the facilitator's account, even though the beneficiary will be Alice's account.

Or am I missing something that prevents this?

Reply:
We can discuss this on call.
I have made the changes, the extra check is not a harm

Reply:
Not sure below is helpful: 

In the contract `msg.sender` is redeemer/staker.  `msg.sender` needs to approve gateway for tokens. 
If X account (staker/ redeemer) owns the token then Y account (facilitator) can't initiate stake and redeem. It can only be done by X.  `TransferFrom` inside gateway happens from `msg.sender`.